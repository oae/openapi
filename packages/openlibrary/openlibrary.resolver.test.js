const { request } = require('graphql-request');
const { gql } = require('@openapi/core/utils');
const { createServer, destroyServer } = require('@openapi/core/testUtils');

const { name: pluginName } = require('./package.json');

let server = null;

beforeAll(async () => {
  server = await createServer([pluginName]);
});

afterAll(async () => {
  await destroyServer(server);
});

const bookFields = gql`
  fragment bookFields on Book {
    title
    description
    createdAt
    lastModifiedAt
    covers {
      small
      medium
      large
    }
    subjectPlaces
    subjectPeople
    subjects
    editions(limit: 2) {
      title
      isbn10
      isbn13
      publishers
      publishDate
      authors {
        name
      }
    }
    authors(limit: 2) {
      name
      createdAt
      lastModifiedAt
      birthDate
      deathDate
      altNames
      personalName
    }
  }
`;

const validBook = {
  title: expect.not.toBeEmpty(),
  description: expect.not.toBeEmpty(),
  createdAt: expect.not.toBeEmpty(),
  lastModifiedAt: expect.not.toBeEmpty(),
  subjectPlaces: expect.toBeArray(),
  subjectPeople: expect.toBeArray(),
  subjects: expect.toBeArray(),
  editions: expect.toBeArray(),
  covers: expect.toBeArray(),
  authors: expect.toBeArray(),
};

describe('openlibrary', () => {
  it(
    'should return books with matching title',
    async () => {
      const query = gql`
        ${bookFields}
        {
          books(title: "Lord of the Rings", limit: 2) {
            ...bookFields
          }
        }
      `;

      const result = await request(server.endpoint, query);
      expect(result.books).toBeArray();
      expect(result.books[0]).toMatchObject(validBook);
    },
    20000
  );

  it(
    'should return books from given author',
    async () => {
      const query = gql`
        ${bookFields}
        {
          booksFromAuthor(author: "tolkien", limit: 2) {
            ...bookFields
          }
        }
      `;

      const result = await request(server.endpoint, query);
      expect(result.booksFromAuthor).toBeArray();
      expect(result.booksFromAuthor[0]).toMatchObject(validBook);
    },
    20000
  );
});
