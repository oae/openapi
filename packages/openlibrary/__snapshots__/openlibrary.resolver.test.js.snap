// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`openlibrary should return books from given author 1`] = `
Object {
  "booksFromAuthor": Array [
    Object {
      "authors": Array [
        Object {
          "altNames": Array [
            "John Ronald Reuel Tolkien",
            "J. R. R. Tolkein",
            "J.R.R. Tolkein",
          ],
          "birthDate": "3 January 1892",
          "createdAt": "2008-04-01T03:28:50.625Z",
          "deathDate": "2 September 1973",
          "lastModifiedAt": "2017-03-31T06:37:59.983Z",
          "name": "J. R. R. Tolkien",
          "personalName": "J. R. R. Tolkien",
        },
      ],
      "covers": Array [
        Object {
          "large": "https://covers.openlibrary.org/w/id/8236263-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/8236263-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/8236263-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/-1-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/-1-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/-1-S.jpg",
        },
      ],
      "createdAt": "2009-10-24T09:28:02.453Z",
      "description": "The classic fantasy novel of a time before the “dominion of men”, this book introduces the reader to Tolkien’s world of hobbits, dwarves, goblins and wizards. The story takes the form of a quest as the hobbit Bilbo Baggins joins Gandalf the wizard Thorin Oakenshield the dwarf on a journey to claim a vast treasure guarded by the evil dragon Smaug. This book is a precursor to Tolkien’s extremely popular Lord of the Rings trilogy, the first of which was published nearly fifteen years later.",
      "editions": Array [
        Object {
          "authors": Array [
            Object {
              "name": "J. R. R. Tolkien",
            },
          ],
          "isbn10": Array [
            "0048230707",
          ],
          "isbn13": null,
          "publishDate": null,
          "publishers": Array [
            "Allen and Unwin",
          ],
          "title": "The Hobbit",
        },
        Object {
          "authors": Array [
            Object {
              "name": "J. R. R. Tolkien",
            },
          ],
          "isbn10": null,
          "isbn13": null,
          "publishDate": null,
          "publishers": Array [
            "Unwin",
          ],
          "title": "The Hobbit",
        },
      ],
      "lastModifiedAt": "2018-09-01T13:46:06.442Z",
      "subjectPeople": Array [
        "Bilbo Baggins",
        "Gandalf the Wizard",
        "J. R. R. Tolkien (1892-1973)",
      ],
      "subjectPlaces": Array [
        "Middle-earth",
      ],
      "subjects": Array [
        "English Fantasy fiction",
        "In library",
        "Middle Earth (Imaginary place)",
        "Accessible book",
        "Bilbo Baggins (Fictitious character)",
        "Fiction",
        "Wizards",
        "Specimens",
        "Juvenile Fantasy fiction",
        "Toy and movable books",
        "Magic in fiction",
        "Open Library Staff Picks",
        "Large type books",
        "juvenile literature",
        "Pictorial works",
        "Magic",
        "Quests (Expeditions)",
        "Graphic novels",
        "Dragons in fiction",
        "Children's stories",
        "Ficción fantástica inglesa",
        "Fantasy in fiction",
        "Translations into Russian",
        "Fantasy fiction",
        "Juvenile fiction",
        "Tierra Media (Lugar imaginario)",
        "Dragons",
        "Fantasy",
        "Protected DAISY",
        "Ficción",
        "Internet Archive Wishlist",
      ],
      "title": "The Hobbit",
    },
    Object {
      "authors": Array [
        Object {
          "altNames": Array [
            "John Ronald Reuel Tolkien",
            "J. R. R. Tolkein",
            "J.R.R. Tolkein",
          ],
          "birthDate": "3 January 1892",
          "createdAt": "2008-04-01T03:28:50.625Z",
          "deathDate": "2 September 1973",
          "lastModifiedAt": "2017-03-31T06:37:59.983Z",
          "name": "J. R. R. Tolkien",
          "personalName": "J. R. R. Tolkien",
        },
      ],
      "covers": Array [
        Object {
          "large": "https://covers.openlibrary.org/w/id/8167232-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/8167232-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/8167232-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/8167229-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/8167229-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/8167229-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/7389663-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/7389663-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/7389663-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/2341290-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/2341290-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/2341290-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/6415824-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/6415824-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/6415824-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/255849-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/255849-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/255849-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/5071064-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/5071064-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/5071064-S.jpg",
        },
      ],
      "createdAt": "2009-10-24T09:28:02.453Z",
      "description": "Frodo and the Companions of the Ring have been beset by danger during their quest to prevent the Ruling Ring from falling into the hands of the Dark Lord by destroying it in the Cracks of Doom. They have lost the wizard, Gandalf, in the battle with an evil spirit in the Mines of Moria; and at the Falls of Rauros, Boromir, seduced by the power of the Ring, tried to seize it by force. While Frodo and Sam made their escape the rest of the company were attacked by Orcs. Now they continue their journey alone down the great River Anduin – alone, that is, save for the mysterious creeping figure that follows wherever they go. J.R.R. Tolkien’s great work of imaginative fiction has been labelled both a heroic romance and a classic fantasy fiction. By turns comic and homely, epic and diabolic, the narrative moves through countless changes of scene and character in an imaginary world which is totally convincing in its detail.",
      "editions": Array [
        Object {
          "authors": Array [
            Object {
              "name": "J. R. R. Tolkien",
            },
          ],
          "isbn10": Array [
            "8445071769",
          ],
          "isbn13": Array [
            "9788445071762",
          ],
          "publishDate": "2001 December",
          "publishers": Array [
            "Minotauro",
          ],
          "title": "Las dos torres",
        },
        Object {
          "authors": Array [],
          "isbn10": Array [
            "6020332276",
          ],
          "isbn13": Array [
            "9786020332277",
          ],
          "publishDate": "2016 August",
          "publishers": Array [
            "Gramedia Pustaka Utama",
          ],
          "title": "The Two Towers",
        },
      ],
      "lastModifiedAt": "2018-08-11T15:51:58.362Z",
      "subjectPeople": Array [
        "Gandalf",
        "Frodo Baggins",
        "Aragorn II",
        "Merry",
        "Pippin",
        "Samwise",
        "Boromir",
        "Saruman the White",
        "Uruk-hai",
        "hobbits",
        "Sauron",
        "Shelob",
        "Legolas",
        "Gimli",
        "Orcs",
        "Huorns",
        "Gollum",
        "Éomer",
        "Gríma Wormtongue",
        "the Balrog",
        "Ents",
        "Treebeard",
        "Rohirrim",
        "Théoden",
      ],
      "subjectPlaces": Array [
        "Tierra Media",
        "River Anduin",
        "Middle Earth",
        "Minas Tirith",
        "Rohan",
        "Fangorn Forrest",
        "Moria",
        "Edoras",
        "Helm's Deep",
        "Isengard",
        "Gondor",
        "Mount Doom",
        "Ithilien",
        "Dead Marshes",
      ],
      "subjects": Array [
        "magic",
        "Ficción fantástica inglesa",
        "Ficción",
        "translations into Spanish",
        "readers for new literates",
        "The Lord of the Rings",
        "English Fantasy fiction",
        "In library",
        "Open Library Staff Picks",
        "Accessible book",
        "Fiction",
        "Large type books",
        "Fantasy",
        "Protected DAISY",
        "Popular Print Disabled Books",
        "Popular Carousel",
        "Internet Archive Wishlist",
      ],
      "title": "The Two Towers",
    },
  ],
}
`;

exports[`openlibrary should return books with matching title 1`] = `
Object {
  "books": Array [
    Object {
      "authors": Array [
        Object {
          "altNames": Array [
            "John Ronald Reuel Tolkien",
            "J. R. R. Tolkein",
            "J.R.R. Tolkein",
          ],
          "birthDate": "3 January 1892",
          "createdAt": "2008-04-01T03:28:50.625Z",
          "deathDate": "2 September 1973",
          "lastModifiedAt": "2017-03-31T06:37:59.983Z",
          "name": "J. R. R. Tolkien",
          "personalName": "J. R. R. Tolkien",
        },
      ],
      "covers": Array [
        Object {
          "large": "https://covers.openlibrary.org/w/id/8069767-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/8069767-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/8069767-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/393994-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/393994-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/393994-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/7382917-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/7382917-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/7382917-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/6308838-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/6308838-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/6308838-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/258027-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/258027-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/258027-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/9630-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/9630-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/9630-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/6943810-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/6943810-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/6943810-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/4248333-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/4248333-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/4248333-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/6409563-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/6409563-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/6409563-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/10447-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/10447-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/10447-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/6673383-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/6673383-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/6673383-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/6420406-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/6420406-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/6420406-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/393435-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/393435-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/393435-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/9632-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/9632-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/9632-S.jpg",
        },
        Object {
          "large": "https://covers.openlibrary.org/w/id/6821986-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/6821986-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/6821986-S.jpg",
        },
      ],
      "createdAt": "2009-10-13T02:46:28.838Z",
      "description": "A one-volume edition of the trilogy which tells of the quest undertaken by the hobbit Frodo and his companions to journey across Middle-earth and cast the evil One Ring, into the Cracks of Doom.",
      "editions": Array [
        Object {
          "authors": Array [
            Object {
              "name": "J. R. R. Tolkien",
            },
          ],
          "isbn10": Array [
            "0345020200",
          ],
          "isbn13": Array [
            "9780345020208",
          ],
          "publishDate": "1970 September",
          "publishers": Array [
            "Ballantine Books",
          ],
          "title": "The Lord of the Rings",
        },
        Object {
          "authors": Array [
            Object {
              "name": "J. R. R. Tolkien",
            },
          ],
          "isbn10": Array [
            "0261103563",
          ],
          "isbn13": Array [
            "9780261103566",
          ],
          "publishDate": "1997 November 3",
          "publishers": Array [
            "HarperCollins Publishers Ltd",
          ],
          "title": "The Hobbit & The Lord of the Rings",
        },
      ],
      "lastModifiedAt": "2018-07-30T23:48:11.201Z",
      "subjectPeople": Array [
        "Aragorn",
        "Frodo",
        "Sam",
        "Pippin",
        "Merry",
        "Gandalf",
        "Shelob",
      ],
      "subjectPlaces": Array [
        "Mordor",
        "Hornburg",
        "Isengard",
        "Cracks of Doom",
        "Fangorn Forest",
      ],
      "subjects": Array [
        "Fiction",
        "Ficción",
        "English Fantasy fiction",
        "Middle Earth (Imaginary place)",
        "Baggins, Frodo (Personaje literario)",
        "Tierra Media (Lugar imaginario)",
        "Frodo Baggins (Fictitious character)",
        "Ficción fantástica inglesa",
        "In library",
        "Gandalf (Fictitious character)",
        "Fantasy fiction",
        "Accessible book",
        "Fiction in English",
        "Frodo Baggins",
        "Open Library Staff Picks",
        "Protected DAISY",
        "Textual Criticism",
        "Internet Archive Wishlist",
      ],
      "title": "The Lord of the Rings",
    },
    Object {
      "authors": Array [
        Object {
          "altNames": null,
          "birthDate": null,
          "createdAt": null,
          "deathDate": null,
          "lastModifiedAt": "2008-04-29T15:03:11.581Z",
          "name": "Cedco Publishing",
          "personalName": null,
        },
      ],
      "covers": Array [
        Object {
          "large": "https://covers.openlibrary.org/w/id/1454705-L.jpg",
          "medium": "https://covers.openlibrary.org/w/id/1454705-M.jpg",
          "small": "https://covers.openlibrary.org/w/id/1454705-S.jpg",
        },
      ],
      "createdAt": "2009-12-10T22:54:38.967Z",
      "description": null,
      "editions": Array [
        Object {
          "authors": Array [
            Object {
              "name": "Cedco Publishing",
            },
          ],
          "isbn10": Array [
            "0768325293",
          ],
          "isbn13": Array [
            "9780768325294",
          ],
          "publishDate": "October 2001",
          "publishers": Array [
            "Cedco Publishing Company",
          ],
          "title": "Lord of the Rings",
        },
        Object {
          "authors": Array [
            Object {
              "name": "Cedco Publishing",
            },
          ],
          "isbn10": Array [
            "0768325242",
          ],
          "isbn13": Array [
            "9780768325249",
          ],
          "publishDate": "October 2001",
          "publishers": Array [
            "Cedco Publishing Company",
          ],
          "title": "Lord of the Rings",
        },
      ],
      "lastModifiedAt": "2010-04-28T09:44:07.374Z",
      "subjectPeople": null,
      "subjectPlaces": null,
      "subjects": null,
      "title": "Lord of the Rings",
    },
  ],
}
`;
